# Stack Overflow Developer Survey 2023 – EDA
This project explores developer demographics, skills, compensation, and AI usage using the Stack Overflow annual survey dataset.

In [1]:
import pandas as pd

## 1. Load Data

In [2]:
df = pd.read_csv("../datasets/survey_results_public.csv")
schema_df = pd.read_csv("../datasets/survey_results_schema.csv")

## 2. Exploring Data

In [3]:
df.shape

(49123, 170)

In [4]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49123 entries, 0 to 49122
Data columns (total 170 columns):
 #    Column                                    Dtype  
---   ------                                    -----  
 0    ResponseId                                int64  
 1    MainBranch                                object 
 2    Age                                       object 
 3    EdLevel                                   object 
 4    Employment                                object 
 5    EmploymentAddl                            object 
 6    WorkExp                                   float64
 7    LearnCodeChoose                           object 
 8    LearnCode                                 object 
 9    LearnCodeAI                               object 
 10   AILearnHow                                object 
 11   YearsCode                                 float64
 12   DevType                                   object 
 13   OrgSize                                   ob

- Checking for duplicated entries

In [5]:
df.loc[df.duplicated()]

Empty DataFrame
Columns: [ResponseId, MainBranch, Age, EdLevel, Employment, EmploymentAddl, WorkExp, LearnCodeChoose, LearnCode, LearnCodeAI, AILearnHow, YearsCode, DevType, OrgSize, ICorPM, RemoteWork, PurchaseInfluence, TechEndorseIntro, TechEndorse_1, TechEndorse_2, TechEndorse_3, TechEndorse_4, TechEndorse_5, TechEndorse_6, TechEndorse_7, TechEndorse_8, TechEndorse_9, TechEndorse_13, TechEndorse_13_TEXT, TechOppose_1, TechOppose_2, TechOppose_3, TechOppose_5, TechOppose_7, TechOppose_9, TechOppose_11, TechOppose_13, TechOppose_16, TechOppose_15, TechOppose_15_TEXT, Industry, JobSatPoints_1, JobSatPoints_4, JobSatPoints_5, JobSatPoints_6, JobSatPoints_7, JobSatPoints_8, JobSatPoints_9, JobSatPoints_10, JobSatPoints_11, JobSatPoints_13, JobSatPoints_14, JobSatPoints_15, JobSatPoints_16, JobSatPoints_15_TEXT, AIThreat, NewRole, ToolCountWork, ToolCountPersonal, Country, Currency, CompTotal, LanguageChoice, LanguageHaveWorkedWith, LanguageWantToWorkWith, LanguageAdmired, LanguagesHaveEntry, LanguagesWantEntry, DatabaseChoice, DatabaseHaveWorkedWith, DatabaseWantToWorkWith, DatabaseAdmired, DatabaseHaveEntry, DatabaseWantEntry, PlatformChoice, PlatformHaveWorkedWith, PlatformWantToWorkWith, PlatformAdmired, PlatformHaveEntry, PlatformWantEntry, WebframeChoice, WebframeHaveWorkedWith, WebframeWantToWorkWith, WebframeAdmired, WebframeHaveEntry, WebframeWantEntry, DevEnvsChoice, DevEnvsHaveWorkedWith, DevEnvsWantToWorkWith, DevEnvsAdmired, DevEnvHaveEntry, DevEnvWantEntry, SOTagsHaveWorkedWith, SOTagsWantToWorkWith, SOTagsAdmired, SOTagsHaveEntry, SOTagsWant Entry, OpSysPersonal use, OpSysProfessional use, OfficeStackAsyncHaveWorkedWith, ...]
Index: []

[0 rows x 170 columns]

## 3. Selecting Relevant Columns
The original dataset contains ~170 columns.  
For this analysis, I focus on the most relevant 33 columns listed below: 

In [6]:
list(df.columns)
columns_needed = ['ResponseId','MainBranch','Age','EdLevel','Employment','WorkExp',
                  'LearnCode','YearsCode','DevType','RemoteWork','Industry','JobSat',
                  'AIThreat','NewRole','Country','Currency','CompTotal','LanguageHaveWorkedWith',
                  'DatabaseHaveWorkedWith','PlatformHaveWorkedWith','WebframeHaveWorkedWith',
                  'DevEnvsHaveWorkedWith','OpSysPersonal use','OpSysProfessional use','OfficeStackAsyncHaveWorkedWith',
                  'CommPlatformHaveWorkedWith','AIModelsChoice','AIModelsHaveWorkedWith','AISelect','AISent',
                  'ConvertedCompYearly','AIToolCurrently partially AI','AIToolCurrently mostly AI'
                  ]

In [7]:
len(columns_needed)

33

In [8]:
df_main = df[columns_needed].copy()

In [9]:
df_main.head(2)

ResponseId                      MainBranch              Age  \
0           1  I am a developer by profession  25-34 years old   
1           2  I am a developer by profession  25-34 years old   

                                           EdLevel Employment  WorkExp  \
0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   Employed      8.0   
1              Associate degree (A.A., A.S., etc.)   Employed      2.0   

                                           LearnCode  YearsCode  \
0  Online Courses or Certification (includes all ...       14.0   
1  Online Courses or Certification (includes all ...       10.0   

               DevType                                         RemoteWork  \
0    Developer, mobile                                             Remote   
1  Developer, back-end  Hybrid (some in-person, leans heavy to flexibi...   

   ...             OpSysProfessional use  \
0  ...  Windows;MacOS;Android;iOS;iPadOS   
1  ...                             MacOS   

                OfficeStackAsyncHaveWorkedWith  \
0  Confluence;GitHub;GitLab;Jira;Markdown File   
1                       Confluence;GitHub;Jira   

                          CommPlatformHaveWorkedWith AIModelsChoice  \
0  Stack Overflow;Reddit;GitHub (public projects,...            Yes   
1  Stack Overflow;Discord;Reddit;Dev.to;Hacker Ne...            Yes   

                              AIModelsHaveWorkedWith  \
0  openAI GPT (chatbot models);openAI Image gener...   
1                        openAI GPT (chatbot models)   

                                      AISelect       AISent  \
0  Yes, I use AI tools monthly or infrequently  Indifferent   
1                   Yes, I use AI tools weekly  Indifferent   

  ConvertedCompYearly                       AIToolCurrently partially AI  \
0             61256.0  Learning about a codebase;Writing code;Search ...   
1            104413.0              Learning new concepts or technologies   

  AIToolCurrently mostly AI  
0                       NaN  
1                       NaN  

[2 rows x 33 columns]

## 4. Missing Values Overview
Before cleaning, let's examine missing values in each column.

In [10]:
df_main.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49123 entries, 0 to 49122
Data columns (total 33 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      49123 non-null  int64  
 1   MainBranch                      49123 non-null  object 
 2   Age                             49123 non-null  object 
 3   EdLevel                         48087 non-null  object 
 4   Employment                      48277 non-null  object 
 5   WorkExp                         42844 non-null  float64
 6   LearnCode                       33516 non-null  object 
 7   YearsCode                       43000 non-null  float64
 8   DevType                         43636 non-null  object 
 9   RemoteWork                      33747 non-null  object 
 10  Industry                        33607 non-null  object 
 11  JobSat                          26643 non-null  float64
 12  AIThreat                        

### Column: EdLevel
- **Type:** Categorical  
- **Missing:** 1036  
- **Action:** Combine `Other` + fill `NaN` as `Others`

In [11]:
df_main['EdLevel'] = df_main['EdLevel'].replace('Other (please specify):', 'Others')

In [12]:
df_main['EdLevel'] = df_main['EdLevel'].fillna('Others')

In [13]:
df_main['EdLevel'].isna().sum()

np.int64(0)

### Column: Employment

- **Type:** Categorical
- **Missing:** 846
- **Action:** Fill `NaN` as `I prefer not to say`** to keep categories consistent and avoid unnecessary assumptions.

In [14]:
df_main['Employment'] = df_main['Employment'].fillna('I prefer not to say')

In [15]:
df_main["Employment"].isna().sum()

np.int64(0)

### Column: WorkExp

- **Type:** Numeric
- **Missing:** 6,279
- **Action:** Filled missing values with the `median`, as the distribution is skewed and median gives a stable central value.

In [16]:
df_main['WorkExp'] = df_main['WorkExp'].fillna(df_main['WorkExp'].median())

In [17]:
df_main['WorkExp'].isna().sum()

np.int64(0)

### Column: LearningMethods
- **Type:** Categorical (multi-select, semicolon-separated)  
- **Missing:** 15607
- **Actions Taken:**  
  - Renamed column from `LearnCode` → `LearningMethods` for clarity.  
  - Replaced `"Other (please specify):"` with `"Others"` to unify the category.  
  - Filled missing values (`NaN`) with `""`.  

In [18]:
df_main.rename(columns={'LearnCode': 'LearningMethods'}, inplace=True)

In [19]:
from collections import Counter
c = Counter()

In [20]:
for lm in df_main['LearningMethods'].dropna().str.split(";"):
    c.update(lm)

In [21]:
my_dict = dict(c)
my_dict

{'Online Courses or Certification (includes all media types)': 10957,
 'Other online resources (e.g. standard search, forum, online community)': 19676,
 'Books / Physical media': 10170,
 'Videos (not associated with specific online course or certification)': 16743,
 'Stack Overflow or Stack Exchange': 17227,
 'Technical documentation (is generated for/by the tool or system)': 22709,
 'AI CodeGen tools or AI-enabled apps': 14732,
 'Colleague or on-the-job training': 8589,
 'Blogs or podcasts': 11672,
 'Coding Bootcamp': 1743,
 'Games or coding challenges': 4042,
 'School (i.e., University, College, etc)': 5555,
 'Other (please specify):': 1068}

In [22]:
df_main["LearningMethods"] = df_main["LearningMethods"].str.replace('Other (please specify):','Others', regex=False)

In [23]:
df_main['LearningMethods'] = df_main['LearningMethods'].fillna('')

### Column: YearsCode
- **Type:** Numerical (float)
- **Missing:** 6123
- **Action:** Filled missing values using the `median`, which is appropriate for skewed numeric distributions and helps avoid distortion from outliers.

In [24]:
df_main['YearsCode'] = df_main['YearsCode'].fillna(df_main['YearsCode'].median())

In [25]:
df_main['YearsCode'].isnull().sum()

np.int64(0)

### Column: DevType

- **Type:** Categorical
- **Missing:** 5487
- **Action:**
    - Replace `Other (please specify):` → `Others`
    - Fill missing values with `Others`

In [26]:
df_main['DevType'] = df_main['DevType'].str.replace('Other (please specify):', 'Others')

In [27]:
df_main['DevType'] = df_main['DevType'].fillna('Others')

In [28]:
df_main['DevType'].isna().sum()

np.int64(0)

### Column: RemoteWork

- **Type:** Categorical  
- **Missing:** Filled using custom category  
- **Actions:** 
    - Standardized hybrid labels: `Hybrid*` → `Hybrid`  
    - Replaced long flexible-work description with `Flexible`  
    - Filled missing values with `Prefer not to say`

In [29]:
df_main['RemoteWork'] = (df_main['RemoteWork']
    .replace(r'Hybrid.*', 'Hybrid', regex=True)
    .replace({
        'Your choice (very flexible, you can come in when you want or just as needed)':'Flexible'
    })
)

In [30]:
df_main['RemoteWork'] = df_main['RemoteWork'].fillna('Prefer not to say')

In [31]:
df_main['RemoteWork'].value_counts(normalize=True)*100

RemoteWork
Prefer not to say    31.301020
Hybrid               25.552185
Remote               22.238055
In-person            12.283452
Flexible              8.625288
Name: proportion, dtype: float64

In [32]:
df_main['RemoteWork'].isna().sum()

np.int64(0)

### Column: Industry

- **Type:** Categorical  
- **Missing:** 15,516  
- **Actions:**  
    - Replaced `Other:` → `Others`  
    - Filled missing values with `Others`

In [33]:
df_main['Industry'] = df_main['Industry'].replace('Other:', 'Others')

In [34]:
df_main['Industry'] = df_main['Industry'].fillna('Others')

In [35]:
df_main['Industry'].isna().sum()

np.int64(0)

### Column: JobSat

- **Type:** Numeric (0–10 satisfaction rating)  
- **Missing:** 22,480  
- **Actions:**  
    - Imputed missing values with the median score (8.0)

In [36]:
df_main['JobSat'] = df_main['JobSat'].fillna(df_main['JobSat'].median())

In [37]:
df_main['JobSat'].isna().sum()

np.int64(0)

### Column: AIThreat

- **Type:** Categorical  
- **Missing:** 13,082  
- **Actions:**  
    - Filled missing values with `I'm not sure` (neutral category)

In [38]:
df_main['AIThreat'] = df_main['AIThreat'].fillna("I'm not sure")

In [39]:
df_main['AIThreat'].isna().sum()

np.int64(0)

### Column: NewRole

- **Type:** Categorical  
- **Missing:** 13,632  
- **Actions:**  
    - Filled missing values with `Prefer not to say`

In [40]:
df_main['NewRole'] = df_main['NewRole'].fillna('Prefer not to say')

In [41]:
df_main['NewRole'].isna().sum()

np.int64(0)

In [42]:
df_main['NewRole'].value_counts()

NewRole
I have neither consider or transitioned into a new career or industry          16192
Prefer not to say                                                              13632
I have somewhat considered changing my career and/or the industry I work in    10219
I have strongly considered changing my career and/or the industry I work in     5249
I have transitioned into a new career and/or industry voluntarily               3107
I have transitioned into a new career and/or industry involuntarily              724
Name: count, dtype: int64

### Column: Country

- **Type:** Categorical (177 unique categories)  
- **Missing:** 13,721  
- **Actions:**  
    - Filled missing values with `Unknown`

In [43]:
df_main['Country'] = df_main['Country'].fillna('Unknown')

In [44]:
df_main['Country'].isna().sum()

np.int64(0)

### Columns: Currency, CompTotal, ConvertedCompYearly

- **Type:** Mixed (categorical + numeric)
- **Actions:**  
    - Verified that `CompTotal` represents annual salary in the respondent’s local currency  
    - Verified that `ConvertedCompYearly` is the same salary converted to USD  
    - Since `ConvertedCompYearly` already provides standardized salary values needed for EDA,  
      the columns `Currency` and `CompTotal` are **dropped** to avoid redundancy  
    - Renamed `ConvertedCompYearly` → `AnnualSalary(USD)` and is **kept** as the primary salary metric for analysis

In [45]:
df_main.loc[:10,['Country','Currency','CompTotal','ConvertedCompYearly']]

Country                  Currency  CompTotal  ConvertedCompYearly
0       Ukraine         EUR European Euro    52800.0              61256.0
1   Netherlands         EUR European Euro    90000.0             104413.0
2       Ukraine     UAH Ukrainian hryvnia  2214000.0              53061.0
3       Ukraine         EUR European Euro    31200.0              36197.0
4       Ukraine  USD United States dollar    60000.0              60000.0
5       Ukraine  USD United States dollar   120000.0             120000.0
6       Ukraine  USD United States dollar     6240.0               6240.0
7       Ukraine  USD United States dollar    72000.0              72000.0
8       Ukraine  USD United States dollar    70000.0              70000.0
9       Ukraine  USD United States dollar     5400.0               5400.0
10      Ukraine     UAH Ukrainian hryvnia  2740980.0              65691.0

In [46]:
df_main.drop(['Currency','CompTotal'], axis=1, inplace=True)

In [47]:
df_main = df_main.rename(columns={'ConvertedCompYearly': 'AnnualSalary(USD)'})

### Column: AnnualSalary(USD)

- **Type:** Numeric (continuous)
- **Missing:** 25,195
- **Actions:**  
    - Salary values are highly skewed and subjective  
    - Imputing (mean/median) would distort distributions and bias salary-related analyses  
    - Therefore **missing values are left as-is**  
    - Rows may be selectively dropped only during analyses that require salary  

In [48]:
df_main['AnnualSalary(USD)'].isna().sum()

np.int64(25195)

In [49]:
df_main['AnnualSalary(USD)'].nunique()

6234

### Column: LanguageHaveWorkedWith

- **Type:** Multi-select categorical
- **Missing:** 17,480  
- **Actions:**  
  - Checked for any default placeholder values using a full frequency counter (e.g., "None", "Default", "N/A"). **None were found**  
  - Missing values represent respondents who did not list any languages → replaced with an **empty string**  
  - Split the column on `;` and cleaned whitespace  
  - Converted each row into a **list of languages**, removing empty entries  
  - This keeps the multi-language structure intact for later EDA (like popularity counts)

In [50]:
langCounter = Counter()

In [51]:
for languages in df_main['LanguageHaveWorkedWith'].dropna().str.split(";"):
    langCounter.update(languages)

lang_dict = dict(langCounter)
langCounter

Counter({'JavaScript': 20986,
         'HTML/CSS': 19681,
         'SQL': 18617,
         'Python': 18388,
         'Bash/Shell (all shells)': 15486,
         'TypeScript': 13852,
         'Java': 9344,
         'C#': 8845,
         'C++': 7479,
         'PowerShell': 7366,
         'C': 6979,
         'PHP': 5990,
         'Go': 5213,
         'Rust': 4717,
         'Kotlin': 3418,
         'Lua': 2906,
         'Assembly': 2244,
         'Ruby': 2046,
         'Dart': 1882,
         'Swift': 1718,
         'R': 1569,
         'Groovy': 1533,
         'Visual Basic (.Net)': 1408,
         'VBA': 1333,
         'MATLAB': 1231,
         'Perl': 1215,
         'GDScript': 1061,
         'Elixir': 847,
         'Scala': 842,
         'Delphi': 796,
         'Lisp': 751,
         'MicroPython': 721,
         'Zig': 678,
         'Erlang': 465,
         'Fortran': 449,
         'Ada': 431,
         'F#': 406,
         'OCaml': 367,
         'Gleam': 353,
         'Prolog': 337,
         'CO

In [52]:
df_main['LanguageHaveWorkedWith'] = (df_main['LanguageHaveWorkedWith']
    .fillna("")
    .str.split(";")
    .apply(lambda x: [i.strip() for i in x if i.strip() != ""])
)

In [53]:
df_main['LanguageHaveWorkedWith'].isna().sum()

np.int64(0)

### Multi-select Columns Cleaned

The following columns all follow the **same cleaning approach**:

- DatabaseHaveWorkedWith  
- PlatformHaveWorkedWith  
- WebframeHaveWorkedWith  
- DevEnvsHaveWorkedWith  
- **renamed** `OpSysPersonal use`→`OpSysPersonal_use` 
- **renamed** `OpSysProfessional use`→`OpSysProfessional_use`
- OfficeStackAsyncHaveWorkedWith  
- CommPlatformHaveWorkedWith  
- AIModelsHaveWorkedWith  
- **renamed** `AIToolCurrently partially AI`→`AIToolCurrently_partially_AI`
- **renamed** `AIToolCurrently mostly AI`→`AIToolCurrently_mostly_AI`  


### Cleaning Steps Applied (Same for All Columns)

- **Checked for default values**  
  - No column contained built-in defaults like “None”, “Prefer not to say”, etc.

- **Standardized all "Other (…)” entries**  
  - Any response starting with `"Other"` was converted to a single category: **"Others"**.

- **Handled missing values**  
  - `NaN` / null values replaced with **empty string `""`** (not "Other" to avoid mixing with real “Other” answers).

- **Split multi-select values into lists**  
  - Each cell originally containing `"A; B; C"` is now converted to:  
    `["A", "B", "C"]`

- **Trimmed whitespace**  
  - Ensured clean, consistent items inside lists.

In [54]:
# List of columns with multi-select values to clean
multi_select_cols = [
    'DatabaseHaveWorkedWith', 
    'PlatformHaveWorkedWith',
    'WebframeHaveWorkedWith', 
    'DevEnvsHaveWorkedWith', 
    'OpSysPersonal use',
    'OpSysProfessional use', 
    'OfficeStackAsyncHaveWorkedWith',
    'CommPlatformHaveWorkedWith',
    'AIModelsHaveWorkedWith',
    'AIToolCurrently partially AI',
    'AIToolCurrently mostly AI'
]

def clean_multiselect_column(col):
    df_main[col] = df_main[col].fillna("")

    df_main[col] = df_main[col].str.replace(r"Other.*", "Others", regex=True)

    df_main[col] = df_main[col].apply(
        lambda x: x.split(";") if ";" in x else (["Others"] if x == "Others" else (""))
    )

In [55]:
for col in multi_select_cols:
    clean_multiselect_column(col)

In [56]:
df_main = df_main.rename(columns={
    'OpSysPersonal use': 'OpSysPersonal_use',
    'OpSysProfessional use': 'OpSysProfessional_use',
    'AIToolCurrently partially AI': 'AIToolCurrently_partially_AI',
    'AIToolCurrently mostly AI': 'AIToolCurrently_mostly_AI'
})

### AI Columns Cleaned

The following columns all follow the **same cleaning approach**:
  
- `AIModelsChoice`→`UsedAIBefore`  
- `AISelect`→`AIToolsUsage`  
- `AISent`→`AISentiment`


### Actions:

- **Handled missing values**  
  - `NaN` / null values replaced with **empty string `""`** (not a default value to avoid making a new category).
- **Trimmed whitespace**  
  - Ensured clean, consistent items inside lists.

In [57]:
ai_cols = ['AIModelsChoice', 'AISelect', 'AISent']
for col in ai_cols:
    df_main[col] = df_main[col].fillna("")
    df_main[col] = df_main[col].str.strip()

In [58]:
df_main = df_main.rename(columns={
    'AIModelsChoice': 'UsedAIBefore',
    'AISelect': 'AIToolsUsage',
    'AISent': 'AISentiment'
})

## 5. Exporting Dataframe to **CSV**

In [59]:
df_main.to_csv("../datasets/cleaned_survey_results.csv",index=False)